perf-cluster-operation/pkg/benchmark.go

```diff
@@ -0,0 +1,142 @@
+	names := make([]string, 0)
+	dirs, err := ioutil.ReadDir(benchmarkRoot)
+	if err != nil {
+		return names, fmt.Errorf("failed to list all benchmarks under %q: %v", benchmarkRoot, err)
+	}
+
+	for _, dir := range dirs {
+		names = append(names, dir.Name())
+	}
+	return names, nil
+}
+
+// benchmarkClusters returns the cluster configs for all benchmarks.
+func benchmarkClusters(repo, benchmarkRoot string) (map[string]ClusterConfig, error) {
+	// clusters is a map of cluster configs
+	// key is the cluster name, value is the cluster config
+	clusters := make(map[string]ClusterConfig)
+	benchmarkNames, err := benchmarkNames(benchmarkRoot)
+	if err != nil {
+		return clusters, err
+	}
+
+	for _, benchmarkName := range benchmarkNames {
+		clusterConfig := clusterConfigForBenchmark(benchmarkRoot, benchmarkName)
+		clusterName := clusterNameForBenchmark(repo, benchmarkName)
+		clusters[clusterName] = clusterConfig
+	}
+
+	return clusters, nil
+}
+
+// clusterConfigForBenchmark returns the cluster config for the given benchmark.
+//
+// Under each benchmark folder, we can put a cluster.yaml file that follows the scheme we define
+// in ClusterConfig struct, in which we specify configuration of the cluster that we use to run the benchmark.
+// If there is no such config file, or the config file is malformed, default config will be used.
+func clusterConfigForBenchmark(benchmarkRoot, benchmarkName string) ClusterConfig {
+	gkeCluster := GKECluster{
+		Config: ClusterConfig{
+			Location:  defaultLocation,
+			NodeCount: defaultNodeCount,
+			NodeType:  defaultNodeType,
+			Addons:    defaultAddons,
+		},
+	}
+
+	configFile := filepath.Join(benchmarkRoot, benchmarkName, clusterConfigFile)
+	if common.FileExists(configFile) {
+		contents, err := ioutil.ReadFile(configFile)
+		if err == nil {
+			if err := yaml.Unmarshal(contents, &gkeCluster); err != nil {
+				log.Printf("Failed to parse the config file %q, default config will be used", configFile)
+			}
+		} else {
+			log.Printf("Failed to read the config file %q, default config will be used", configFile)
+		}
+	}
+
+	return gkeCluster.Config
+}
+
+// clusterNameForBenchmark prepends repo name to the benchmark name, and use it as the cluster name.
+func clusterNameForBenchmark(benchmarkName, repo string) string {
+	return fmt.Sprintf("%s-%s", repo, benchmarkName)
+}
+
+// benchmarkNameForCluster removes repo name prefix from the cluster name, to get the real benchmark name.
+// If the cluster does not belong to the given repo, return an empty string.
+func benchmarkNameForCluster(clusterName, repo string) string {
+	if !clusterBelongsToRepo(clusterName, repo) {
+		return ""
+	}
+	return strings.TrimPrefix(clusterName, repo+"-")
+}
+
+// clusterBelongsToRepo determines if the cluster belongs to the repo, by checking if it has the repo prefix.
+func clusterBelongsToRepo(clusterName, repo string) bool {
+	return strings.HasPrefix(clusterName, repo+"-")
```

Comment:
As I mentioned in the following comment, it is indeed a problem for `eventing` and `eventing-contrib`, but there is a workaround by using two different projects for them.

If we really want to differentiate them, we'll need a hack way, since cluster name only allows `lowercase letters, numbers, or hyphens`. WDYT?

Reply:
can we check it from the end? Hopefully there is no `-` in your benchmark names

Reply:
There will be `-` in the benchmark names. But adding as the postfix can indeed solve the problem, since there can be `eventing-contrib`, but I believe there will be less opportunity to have `contrib-eventing`...
I can make the change if you are also fine with postfix.

Reply:
Not sure how double `-` can solve this problem, let's sync up offline.

Reply:
Done with changing to double `-`